lib/google/gax/api_call.rb

```diff
@@ -45,74 +46,63 @@ class ApiCall
       #
-      def initialize stub_method, timeout: nil, metadata: nil, retry_policy: nil, params_extractor: nil,
-                     exception_transformer: nil
-        @stub_method           = stub_method
-        @timeout               = timeout
-        @metadata              = metadata
-        @retry_policy          = retry_policy
-        @params_extractor      = params_extractor
-        @exception_transformer = exception_transformer
+      def initialize stub_method
+        @stub_method = stub_method
       end
 
       ##
       # Invoke the API call.
       #
       # @param request [Object] The request object.
-      # @param options [CallOption, Hash] The options for making the API call.
-      # @param block [Proc] The proc to call when the API call is made.
+      # @param options [ApiCall::Options, Hash] The options for making the API call. A Hash can be provided to customize
+      #   the options object, using keys that match the arguments for {ApiCall::Options.new}. This object should only be
+      #   used once.
+      # @param on_response [Proc] A Proc object to format the response object. The Proc should accept response as an
+      #   argument, and return a formatted response object. Optional.
+      # @param on_operation [Proc] A Proc object to provide a callback of the response and operation objects. The Proc
+      #   will be called with both the response and operation objects. Optional.
+      # @param on_stream [Proc] A Proc object to provide a callback for every streamed response received. The Proc will
+      #   be called with the response object. Should only be used on Bidi and Server streaming RPC calls. Optional.
       #
-      def call request, options: nil, &block
-        options = init_call_options options
-
-        apply_params_extractor! request, options
-
+      # @return [Object, Thread] The response object. Or, when `on_stream` is provided, a thread running the callback
+      #   on streamed response is returned.
+      #
+      def call request, options: nil, on_response: nil, on_operation: nil, on_stream: nil
+        # Converts hash and nil to an options object
+        options = ApiCall::Options.new options.to_h if options.respond_to? :to_h
+        block = compose_stream_proc on_stream: on_stream, on_response: on_response
         deadline = calculate_deadline options
 
         begin
-          op = @stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true
-
-          res = op.execute
-          yield op if block
-          res
+          operation = stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true, &block
+
+          if on_stream
+            Thread.new { operation.execute }
+          else
+            response = operation.execute
+            response = on_response.call response if on_response
+            on_operation&.call response, operation
```

Comment:
Do we need both `on_response` and `on_operation` (i.e. vs just ignoring the parameter)?

Reply:
Yes, they perform two different behaviors, and are instantiated by different actions. `on_response` formats the response, and is used for LRO and Paging, and `on_operation` is for peeking at trailing metadata.

If a client method needs to wrap the response, it will send a `on_response` proc. If a user wants to peek at trailing metadata they pass a block to the client, which will be sent as the `on_operation` proc.

Reply:
Ah, ok. So I was more confused than I thought...

Looking through the whole method again, and ignoring the names, what you're doing seems fine. The naming I find quite counter-intuitive though, particularly `response` / `on_response` (which can mean either the raw or formatted response) and `on_operation` (which sounds like LRO but it's not).

I am not great at names, but something like this would be clearer to me:
`on_response` -> `format_response` or `on_raw_response` or `on_grpc_response` or `on_data`, etc.
`on_operation` -> `on_complete` or `on_done`, `on_result`, etc.
`on_stream` -> this one is clear as-is. Perhaps `on_next_result` or similar for consistency depending on the other names used.

From a lifecycle perspective, I like this:
`on_data` -> `on_done` (unary)
`on_data` -> `on_stream` -> `on_data` -> `on_stream` -> `...` ->`on_done` (streaming)
rather than this, which sounds backwards to me:
`on_response` -> `on_operation`

Reply:
What do you think about the names `format_response`, `operation_callback`, and `stream_callback`?

Reply:
I like that and I think it's much clearer. Is `*_callback` a more common convention than `on_*`? I have no preference and leave it up to you - just curious.

It's probably worth mentioning in the doc comments that the `operation` is a grpc operation (maybe reference the type?) so it's not confused with the LRO concept that we also have floating around in this lib.